doxygen/man/man3/ClientModel.3
.TH "ClientModel" 3 "Wed Feb 10 2016" "Version 1.0.0.0" "darksilk" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ClientModel \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <clientmodel\&.h>\fP
.PP
Inherits QObject\&.
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBupdateTimer\fP ()"
.br
.ti -1c
.RI "void \fBupdateSnTimer\fP ()"
.br
.ti -1c
.RI "void \fBupdateNumConnections\fP (int numConnections)"
.br
.ti -1c
.RI "void \fBupdateAlert\fP (const QString &\fBhash\fP, int status)"
.br
.in -1c
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBnumConnectionsChanged\fP (int count)"
.br
.ti -1c
.RI "void \fBnumBlocksChanged\fP (int count)"
.br
.ti -1c
.RI "void \fBstrStormnodesChanged\fP (const QString &strStormnodes)"
.br
.ti -1c
.RI "void \fBalertsChanged\fP (const QString &warnings)"
.br
.ti -1c
.RI "void \fBbytesChanged\fP (quint64 totalBytesIn, quint64 totalBytesOut)"
.br
.ti -1c
.RI "void \fBmessage\fP (const QString &title, const QString &message, bool modal, unsigned int style)"
.br
.RI "\fIAsynchronous message notification\&. \fP"
.ti -1c
.RI "void \fBshowProgress\fP (const QString &title, int nProgress)"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBClientModel\fP (\fBOptionsModel\fP *optionsModel, QObject *parent=0)"
.br
.ti -1c
.RI "\fB~ClientModel\fP ()"
.br
.ti -1c
.RI "\fBOptionsModel\fP * \fBgetOptionsModel\fP ()"
.br
.ti -1c
.RI "\fBPeerTableModel\fP * \fBgetPeerTableModel\fP ()"
.br
.ti -1c
.RI "int \fBgetNumConnections\fP () const "
.br
.ti -1c
.RI "QString \fBgetStormnodeCountString\fP () const "
.br
.ti -1c
.RI "int \fBgetNumBlocks\fP () const "
.br
.ti -1c
.RI "int \fBgetNumBlocksAtStartup\fP ()"
.br
.ti -1c
.RI "quint64 \fBgetTotalBytesRecv\fP () const "
.br
.ti -1c
.RI "quint64 \fBgetTotalBytesSent\fP () const "
.br
.ti -1c
.RI "QDateTime \fBgetLastBlockDate\fP () const "
.br
.ti -1c
.RI "bool \fBisTestNet\fP () const "
.br
.RI "\fIReturn true if client connected to testnet\&. \fP"
.ti -1c
.RI "bool \fBinInitialBlockDownload\fP () const "
.br
.RI "\fIReturn true if core is doing initial block download\&. \fP"
.ti -1c
.RI "bool \fBisImporting\fP () const "
.br
.RI "\fIReturn true if core is importing blocks\&. \fP"
.ti -1c
.RI "QString \fBgetStatusBarWarnings\fP () const "
.br
.RI "\fIReturn warnings to be displayed in status bar\&. \fP"
.ti -1c
.RI "QString \fBformatFullVersion\fP () const "
.br
.ti -1c
.RI "QString \fBformatBuildDate\fP () const "
.br
.ti -1c
.RI "bool \fBisReleaseVersion\fP () const "
.br
.ti -1c
.RI "QString \fBclientName\fP () const "
.br
.ti -1c
.RI "QString \fBformatClientStartupTime\fP () const "
.br
.in -1c
.SH "Detailed Description"
.PP 
Model for DarkSilk network client\&. 
.PP
Definition at line 18 of file clientmodel\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "ClientModel::ClientModel (\fBOptionsModel\fP * optionsModel, QObject * parent = \fC0\fP)\fC [explicit]\fP"

.PP
Definition at line 27 of file clientmodel\&.cpp\&.
.SS "ClientModel::~ClientModel ()"

.PP
Definition at line 49 of file clientmodel\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void ClientModel::alertsChanged (const QString & warnings)\fC [signal]\fP"

.SS "void ClientModel::bytesChanged (quint64 totalBytesIn, quint64 totalBytesOut)\fC [signal]\fP"

.SS "QString ClientModel::clientName () const"

.PP
Definition at line 253 of file clientmodel\&.cpp\&.
.SS "QString ClientModel::formatBuildDate () const"

.PP
Definition at line 243 of file clientmodel\&.cpp\&.
.SS "QString ClientModel::formatClientStartupTime () const"

.PP
Definition at line 258 of file clientmodel\&.cpp\&.
.SS "QString ClientModel::formatFullVersion () const"

.PP
Definition at line 238 of file clientmodel\&.cpp\&.
.SS "QDateTime ClientModel::getLastBlockDate () const"

.PP
Definition at line 86 of file clientmodel\&.cpp\&.
.SS "int ClientModel::getNumBlocks () const"

.PP
Definition at line 64 of file clientmodel\&.cpp\&.
.SS "int ClientModel::getNumBlocksAtStartup ()"

.PP
Definition at line 70 of file clientmodel\&.cpp\&.
.SS "int ClientModel::getNumConnections () const"

.PP
Definition at line 54 of file clientmodel\&.cpp\&.
.SS "\fBOptionsModel\fP * ClientModel::getOptionsModel ()"

.PP
Definition at line 178 of file clientmodel\&.cpp\&.
.SS "\fBPeerTableModel\fP * ClientModel::getPeerTableModel ()"

.PP
Definition at line 183 of file clientmodel\&.cpp\&.
.SS "QString ClientModel::getStatusBarWarnings () const"

.PP
Return warnings to be displayed in status bar\&. 
.PP
Definition at line 173 of file clientmodel\&.cpp\&.
.SS "QString ClientModel::getStormnodeCountString () const"

.PP
Definition at line 59 of file clientmodel\&.cpp\&.
.SS "quint64 ClientModel::getTotalBytesRecv () const"

.PP
Definition at line 76 of file clientmodel\&.cpp\&.
.SS "quint64 ClientModel::getTotalBytesSent () const"

.PP
Definition at line 81 of file clientmodel\&.cpp\&.
.SS "bool ClientModel::inInitialBlockDownload () const"

.PP
Return true if core is doing initial block download\&. 
.PP
Definition at line 163 of file clientmodel\&.cpp\&.
.SS "bool ClientModel::isImporting () const"

.PP
Return true if core is importing blocks\&. 
.PP
Definition at line 168 of file clientmodel\&.cpp\&.
.SS "bool ClientModel::isReleaseVersion () const"

.PP
Definition at line 248 of file clientmodel\&.cpp\&.
.SS "bool ClientModel::isTestNet () const"

.PP
Return true if client connected to testnet\&. 
.PP
Definition at line 158 of file clientmodel\&.cpp\&.
.SS "void ClientModel::message (const QString & title, const QString & message, bool modal, unsigned int style)\fC [signal]\fP"

.PP
Asynchronous message notification\&. 
.SS "void ClientModel::numBlocksChanged (int count)\fC [signal]\fP"

.SS "void ClientModel::numConnectionsChanged (int count)\fC [signal]\fP"

.SS "void ClientModel::showProgress (const QString & title, int nProgress)\fC [signal]\fP"

.SS "void ClientModel::strStormnodesChanged (const QString & strStormnodes)\fC [signal]\fP"

.SS "void ClientModel::updateAlert (const QString & hash, int status)\fC [slot]\fP"

.PP
Definition at line 141 of file clientmodel\&.cpp\&.
.SS "void ClientModel::updateNumConnections (int numConnections)\fC [slot]\fP"

.PP
Definition at line 136 of file clientmodel\&.cpp\&.
.SS "void ClientModel::updateSnTimer ()\fC [slot]\fP"

.PP
Definition at line 118 of file clientmodel\&.cpp\&.
.SS "void ClientModel::updateTimer ()\fC [slot]\fP"

.PP
Definition at line 96 of file clientmodel\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for darksilk from the source code\&.


CHANGELOG
Base58 Change, Port Changes and Minor Fixes
Addition of OBJ directory for Daemon builds
Version updated
ReadMe updated
Instamine code for Silk/Dark weaver swap/burn added to main.cpp
OpenSSL update
LevelDB 1.18 Update
Fix to guard against OpenSSL's new strict DER checks, fixed getbalance and other minor fixes
Spelling edits
Enabled locking of Stormnodes via the Stormnode configuration file
Unset Sandstorm checkbox if there were non-anonymized inputs selected in coincontrol, give warning
Removal of target space change and hardfork
Fixed time to switch "out of sync" messages
Fix twice denominatedBalance calculations
Solve performance problems by caching and fix probable security issue
RPC: Allow "sandstorm" only with ENABLE_WALLET
Disconnect Sandstorm timer on Overview page destruction
Fix spaces/monospace fonts in qt console
Mac compatibility so users can switch tabs using CMD+(NUM)
Added better instructions to get libsecp256k1 installed in case of library issues within ReadMe
Added secp256k1 libs to /deps in case for any reason the libs are not available from github anymore
Increased maximum block size to 50MB (50x Bitcoin Core)
Added safe shutdown for Windows
Fixed sync issues and CPU usage
Add event support for Tor
rpc: Prevent easy memory exhaustion attack 
Fixed issue where using setaccount with a foreign address causes the address to be added to your receiving addresses.
Fix priority calculation in CreateTransaction
darksilk.conf added to /src
GUI Styling changes
Statistics Page Added and Icons Updated
Clickable Padlock to Unlock Wallet and Unlock for Staking Only
Fixed GUI bugs
Assertion Error Fix for MacOSX
Sandstorm GUI Fixes
Decentralised DarkSilkMarket added
Overview Transactions UI Fixes
Messaging, GUI Update, Null Pointer Fix, Various Fixes/Changes
Windows Build Fix
Addition of possibly needed dependency
Stormnode Collateral Increase
Marketplace RPC Calls | Cleanup
Amended Spork Max Value for Stormnode Collateral
Removal of unused Spork
Minimum Transaction fee changed to 0.0001DRKSLK
Add support for miniupnpc API version 14
DARKSILK_TIMEDATA_MAX_SAMPLES 200
Add time offset to getpeerinfo output
Don't check duplicate stake while importing
Ignore blocks received while importing
Further Market RPC JSON calls
marketbuyrequest RPC command
marketmybuys RPC call for market
MacOSX Fixes and GUI Fix
Market RPC Minor Fixes
darksilk-qt.pro cleanup
DarkSilk-qt.app renamed to DarkSilk.app for OSX
Replaced stake min age with min confirmation
Limited drop of difficulty per block
Made special marker non spendable
Added nonspendable encoding for block signing key
BIP66 compatibility
Replaced NOP7 with CHECKLOCKTIMEVERIFY
Enforce new script verification flags
Static pos reward
Demand canonical block sig from updated peers
Submitblock: Add submission of stake template
Reward and Min Conf Blocks Changed
Checkpointing Removal
Gitian MacDeploy Graphics Added
IsProtocolV1/IsProtocolV2 hard fork switches removed
FutureDriftV1/FutureDriftV2 reverted back to FutureDrift
CheckStakeKernelHashV1/CheckStakeKernalHashV2 reverted back to CheckStakeKernelHash
Unsigned Int GetTargetSpacing changed to Static Const Int POS_TARGET_SPACING
Fix boost::get usage with Boost 1.58 in rpcrawtransaction.cpp 
Static PoS Reward of 1 DRKSLK
Prevent redefinition compiler warning
Fix issue where getrawtransaction RPC call does not work
PoW additions to GetNextTargetRequired
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
LevelDB windows cross compile missing files
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
Removal of Dead Code
Fix for building with Boost 1.58
LevelDB Cross compile fixes
DarkSilk-Qt: cleanup / optimise addressbookpage
Discourage fee sniping with nLockTime
Removal of dead code from main.h
Changed ProofOfStakeReward to use static const
Diff changes
Shadowchat Messaging Updates
DNSSeed Thread Fix
Stormnode payments start on block 82002
Restore hex to getrawtransaction vout scriptPubkey
static const rename | nActualSpacing code reform
Removal of Keepass until later date that this doesn't cause an issue with Windows builds
Updated dependencies, added Gitian OSX files
Remove unused GetProofOfStakeReward
Faster Retargetting for PoW
Fixed PoS target spacing
Fix payee selectoin on StormnodePayments::ProcessBlock
fix log output / more notCapableReason-s
HasCollateralInputs should check only for at least 1 collateral not 2
Track lastTimeChanged for mnodeman and ask for update only if UpdateNeeded
Added sstx support for free transactions 
Fix for src/qt/stormnodemanager.cpp
RPC Fixes
Fix DoAutomaticDenominating
Few Fixes for Processing Extra Messages
Reversioning and restriction change
Check and remove expired Stormnodes on Client start
Maps in snodeman
bool warning fix
SSEG Update
Show new number of SN's on add/remove in logs | Move output to debug level
sigTime
Temp Fix prior to major changes
SPORK Additions/Changes | SANDSTORM fee removal
prevent ganing sigTime too much on initial ssee
SandStorm/InstantX Functionality | Send with SandStorm on sendcoinsdialog.ui
Fix Filtering for active | add spaces to table-fy console view | add spaces to IP only in full mode
Merged Stormnode blinding
fix compiling for windows
fix console spam
Make stormnodes.dat verification more specific, prevent from overwriting corruped
files with unknown format, give more info in debug.log output
fixed shared key usage
fix even more console spam
rename stormnodes.dat->mncache.dat / fix strings
save nsqCount in mncache.dat
[Qt] add more NULL pointer checks in darksilkgui.cpp
close wallet regardless of "minimize to tray" settings
Added coincontrol sandstorm rounds column
Stormnode Proof of Service Added
Add structure for donating a percentage of ones earnings to another party via the protocol
Fixed snw message issue
fix pubkey/donation conditions
Fixed blinding submission process
increasing ss timeouts
fixed ssee sync
fixed Stormnode payments FindOldestNotInVec

src/qt/clientmodel.cpp

#include <QDateTime>
#include <QTimer>
#include <QDebug>

#include "clientmodel.h"
#include "guiconstants.h"
#include "peertablemodel.h"
#include "optionsmodel.h"
#include "addresstablemodel.h"
#include "transactiontablemodel.h"
#include "chainparams.h"
#include "alert.h"
#include "main.h"
#include "ui_interface.h"
#include "stormnodeman.h"
#include "stormnode-sync.h"

#ifdef USE_NATIVE_I2P
#include "i2p.h"
#endif

static const int64_t nClientStartupTime = GetTime();

ClientModel::ClientModel(OptionsModel *optionsModel, QObject *parent) :
    QObject(parent),
    optionsModel(optionsModel),
    peerTableModel(0),
    cachedNumBlocks(0),
    cachedStormnodeCountString(""),
    numBlocksAtStartup(-1),
    pollTimer(0)
{
    peerTableModel = new PeerTableModel(this);
    pollTimer = new QTimer(this);
    connect(pollTimer, SIGNAL(timeout()), this, SLOT(updateTimer()));
    pollTimer->start(MODEL_UPDATE_DELAY);

    pollSnTimer = new QTimer(this);
    connect(pollSnTimer, SIGNAL(timeout()), this, SLOT(updateSnTimer()));
    // no need to update as frequent as data for balances/txes/blocks
    pollSnTimer->start(MODEL_UPDATE_DELAY * 4);

    subscribeToCoreSignals();
}

ClientModel::~ClientModel()
{
    unsubscribeFromCoreSignals();
}

int ClientModel::getNumConnections() const
{
    return vNodes.size();
}

QString ClientModel::getStormnodeCountString() const
{
    return QString::number((int)snodeman.CountEnabled()) + " / " + QString::number((int)snodeman.size());
}

int ClientModel::getNumBlocks() const
{
    LOCK(cs_main);
    return nBestHeight;
}

int ClientModel::getNumBlocksAtStartup()
{
    if (numBlocksAtStartup == -1) numBlocksAtStartup = getNumBlocks();
    return numBlocksAtStartup;
}

quint64 ClientModel::getTotalBytesRecv() const
{
    return CNode::GetTotalBytesRecv();
}

quint64 ClientModel::getTotalBytesSent() const
{
    return CNode::GetTotalBytesSent();
}

QDateTime ClientModel::getLastBlockDate() const
{
    LOCK(cs_main);
    if (pindexBest)
        return QDateTime::fromTime_t(pindexBest->GetBlockTime());
    else
        return QDateTime::fromTime_t(Params().GenesisBlock().nTime); // Genesis block's time of current network
}


void ClientModel::updateTimer()
{
    // Get required lock upfront. This avoids the GUI from getting stuck on
    // periodical polls if the core is holding the locks for a longer time -
    // for example, during a wallet rescan.
    TRY_LOCK(cs_main, lockMain);
    if(!lockMain)
        return;
    // Some quantities (such as number of blocks) change so fast that we don't want to be notified for each change.
    // Periodically check and update with a timer.
    int newNumBlocks = getNumBlocks();

    if(cachedNumBlocks != newNumBlocks)
    {
        cachedNumBlocks = newNumBlocks;

        emit numBlocksChanged(newNumBlocks);
    }

    emit bytesChanged(getTotalBytesRecv(), getTotalBytesSent());
}

void ClientModel::updateSnTimer()
{
    // Get required lock upfront. This avoids the GUI from getting stuck on
    // periodical polls if the core is holding the locks for a longer time -
    // for example, during a wallet rescan.
    TRY_LOCK(cs_main, lockMain);
    if(!lockMain)
        return;
    QString newStormnodeCountString = getStormnodeCountString();

    if (cachedStormnodeCountString != newStormnodeCountString)
    {
        cachedStormnodeCountString = newStormnodeCountString;

        emit strStormnodesChanged(cachedStormnodeCountString);
    }
}

void ClientModel::updateNumConnections(int numConnections)
{
    emit numConnectionsChanged(numConnections);
}

void ClientModel::updateAlert(const QString &hash, int status)
{
    // Show error message notification for new alert
    if(status == CT_NEW)
    {
        uint256 hash_256;
        hash_256.SetHex(hash.toStdString());
        CAlert alert = CAlert::getAlertByHash(hash_256);
        if(!alert.IsNull())
        {
            emit message(tr("Network Alert"), QString::fromStdString(alert.strStatusBar), false, CClientUIInterface::ICON_ERROR);
        }
    }

    emit alertsChanged(getStatusBarWarnings());
}

bool ClientModel::isTestNet() const
{
    return TestNet();
}

bool ClientModel::inInitialBlockDownload() const
{
    return IsInitialBlockDownload();
}

bool ClientModel::isImporting() const
{
    return fImporting;
}

QString ClientModel::getStatusBarWarnings() const
{
    return QString::fromStdString(GetWarnings("statusbar"));
}

OptionsModel *ClientModel::getOptionsModel()
{
    return optionsModel;
}

PeerTableModel *ClientModel::getPeerTableModel()
{
    return peerTableModel;
}

#ifdef USE_NATIVE_I2P
QString ClientModel::formatI2PNativeFullVersion() const
{
    return QString::fromStdString(FormatI2PNativeFullVersion());
}

void ClientModel::updateNumI2PConnections(int numI2PConnections)
{
    emit numI2PConnectionsChanged(numI2PConnections);
}

int ClientModel::getNumI2PConnections() const
{
    return nI2PNodeCount;
}

QString ClientModel::getPublicI2PKey() const
{
    return QString::fromStdString(I2PSession::Instance().getMyDestination().pub);
}

QString ClientModel::getPrivateI2PKey() const
{
    return QString::fromStdString(I2PSession::Instance().getMyDestination().priv);
}

bool ClientModel::isI2PAddressGenerated() const
{
    return I2PSession::Instance().getMyDestination().isGenerated;
}

bool ClientModel::isI2POnly() const
{
    return IsI2POnly();
}

QString ClientModel::getB32Address(const QString& destination) const
{
    return QString::fromStdString(I2PSession::GenerateB32AddressFromDestination(destination.toStdString()));
}

void ClientModel::generateI2PDestination(QString& pub, QString& priv) const
{
    const SAM::FullDestination generatedDest = I2PSession::Instance().destGenerate();
    pub = QString::fromStdString(generatedDest.pub);
    priv = QString::fromStdString(generatedDest.priv);
}

#endif

QString ClientModel::formatFullVersion() const
{
    return QString::fromStdString(FormatFullVersion());
}

QString ClientModel::formatBuildDate() const
{
    return QString::fromStdString(CLIENT_DATE);
}

bool ClientModel::isReleaseVersion() const
{
    return CLIENT_VERSION_IS_RELEASE;
}

QString ClientModel::clientName() const
{
    return QString::fromStdString(CLIENT_NAME);
}

QString ClientModel::formatClientStartupTime() const
{
    return QDateTime::fromTime_t(nClientStartupTime).toString();
}

// Handlers for core signals
static void ShowProgress(ClientModel *clientmodel, const std::string &title, int nProgress)
{
    // emits signal "showProgress"
    QMetaObject::invokeMethod(clientmodel, "showProgress", Qt::QueuedConnection,
                              Q_ARG(QString, QString::fromStdString(title)),
                              Q_ARG(int, nProgress));
}

static void NotifyNumConnectionsChanged(ClientModel *clientmodel, int newNumConnections)
{
    // Too noisy: qDebug() << "NotifyNumConnectionsChanged : " + QString::number(newNumConnections);
    QMetaObject::invokeMethod(clientmodel, "updateNumConnections", Qt::QueuedConnection,
                              Q_ARG(int, newNumConnections));
}

#ifdef USE_NATIVE_I2P
static void NotifyNumI2PConnectionsChanged(ClientModel *clientmodel, int newNumI2PConnections)
{
    QMetaObject::invokeMethod(clientmodel, "updateNumI2PConnections", Qt::QueuedConnection,
                              Q_ARG(int, newNumI2PConnections));
}
#endif

static void NotifyAlertChanged(ClientModel *clientmodel, const uint256 &hash, ChangeType status)
{
    qDebug() << "NotifyAlertChanged : " + QString::fromStdString(hash.GetHex()) + " status=" + QString::number(status);
    QMetaObject::invokeMethod(clientmodel, "updateAlert", Qt::QueuedConnection,
                              Q_ARG(QString, QString::fromStdString(hash.GetHex())),
                              Q_ARG(int, status));
}

void ClientModel::subscribeToCoreSignals()
{
    // Connect signals to client
    uiInterface.ShowProgress.connect(boost::bind(ShowProgress, this, _1, _2));
    uiInterface.NotifyNumConnectionsChanged.connect(boost::bind(NotifyNumConnectionsChanged, this, _1));
    uiInterface.NotifyAlertChanged.connect(boost::bind(NotifyAlertChanged, this, _1, _2));
#ifdef USE_NATIVE_I2P
    uiInterface.NotifyNumI2PConnectionsChanged.connect(boost::bind(NotifyNumI2PConnectionsChanged, this, _1));
#endif
}

void ClientModel::unsubscribeFromCoreSignals()
{
    // Disconnect signals from client
    uiInterface.ShowProgress.disconnect(boost::bind(ShowProgress, this, _1, _2));
    uiInterface.NotifyNumConnectionsChanged.disconnect(boost::bind(NotifyNumConnectionsChanged, this, _1));
    uiInterface.NotifyAlertChanged.disconnect(boost::bind(NotifyAlertChanged, this, _1, _2));
#ifdef USE_NATIVE_I2P
    uiInterface.NotifyNumI2PConnectionsChanged.disconnect(boost::bind(NotifyNumI2PConnectionsChanged, this, _1));
#endif
}
